scripts/Muon/GUI/Common/plot_widget/plotting_canvas/plotting_canvas_presenter.py

```diff
@@ -44,31 +43,24 @@ def plot_workspaces(self, workspace_names: List[str], workspace_indices: List[in
         workspace_plot_info = self._model.create_workspace_plot_information(workspace_names, workspace_indices,
                                                                             self._options_presenter.get_errors())
         if not hold_on:
-            # Remove data which is currently plotted and not in the new workspace_plot_info
-            workspaces_info_to_remove = [plot_info for plot_info in self._view.plotted_workspace_information
-                                         if plot_info not in workspace_plot_info]
-            self._view.remove_workspace_info_from_plot(workspaces_info_to_remove)
+            self._view.clear_all_workspaces_from_plot()
 
         # Add workspace info which is currently not plotted
-        workspace_info_to_add = [plot_info for plot_info in workspace_plot_info if plot_info
-                                 not in self._view.plotted_workspace_information]
+        workspace_info_to_add = [plot_info for plot_info in workspace_plot_info if plot_info]
         self._view.add_workspaces_to_plot(workspace_info_to_add)
         self._set_axes_limits_and_titles(autoscale)
 
     def remove_workspace_names_from_plot(self, workspace_names: List[str]):
```

Comment:
I understand that the colour resetting is a bit jarring on the eyes, but the intention here was not to redraw a full plot every time you wanted to remove a line. For instance if youre analysing 5 runs, with 4 groups per run and you remove 1 group from the analysis you end up redrawing 16 lines for no purpose.

Reply:
My understanding is that the speed hit from redrawing plots comes from when you call canvas.draw. We are still only calling this once here so I don;t expect a big cost from re-doing all the lines. If this is not the case and we are seeing GUI unresponsives/ delays from with this change then I agree that a different solution is needed though.

Reply:
I think I have just spotted a logic error though where with hold-on you can add the same line to the plot multiple times.

Reply:
Do you know if the speed of the re-draw scales with how many lines are changed before you trigger it? I had assumed not based on the fact that calling redraw on a plot which has not changed is still expensive implying that matplotlib re-draws the entire plot each time anyway.

Reply:
This is worth looking into, I'll do some quick tests and get back to you.

Reply:
Yeah this may be true, worth looking into - I'm not entirely sure on the answer. I guess if it redraws all the plot anyway then it doesn't matter, outside of needing to recreate all the artist objects, but not sure how expensive that is anyway.

Reply:
I've written the script below as a quick test of what the timings are here. My results are:

Initial plot creation time 0.38223910331726074
Time to redraw with full reset 0.3211688995361328
Full reset just redraw time 0.27391624450683594
Remove redraw costs 0.27213406562805176
Just redraw costs 0.27361178398132324
Second Just redraw costs 0.3972156047821045
Third Just redraw costs 0.3894617557525635
Fourth Just redraw costs 0.3889446258544922
Reset all axes redraw costs 0.5951526165008545
Reset all axes redraw only costs 0.40745043754577637

So doing a full reset on an axis is slightly slower but not enormously so compared with removing a single line. This discrepancy will increase with the number of lines on an axes though but I can't see a way around a slight speed cost if we want to re-do the line colors. What is costly is resetting axes you don't have to so for tiled plots it may be worth trying to be a bit cleverer and only clearing affected axes.

I also noticed some slight oddities where successive redraws were reliably longer than the first redraw and that re-running the script without closing the figures made every step take longer the second time round so would be interested to see whether you saw this as well.

From a problem space perspective I am wondering whether we need to re-color in line removal or not. The downside is that when a re-plot is triggered the lines colors will change anyway defaulting to the first x colours in the colour cycle so feels least confusing do to the re-colouring here.


The script is below, you will need to load MUSR62260-3 in the Muon Analysis GUI to get the workspaces it expects in the ADS.
```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

import matplotlib.pyplot as plt
from mantid.api import AnalysisDataService as ADS
import time

MUSR62261__Group__top__Asymmetry__MA = ADS.retrieve('MUSR62261; Group; top; Asymmetry; MA')
MUSR62261__Group__bottom__Asymmetry__MA = ADS.retrieve('MUSR62261; Group; bottom; Asymmetry; MA')
MUSR62262__Group__bkwd__Asymmetry__MA = ADS.retrieve('MUSR62262; Group; bkwd; Asymmetry; MA')
MUSR62262__Group__top__Asymmetry__MA = ADS.retrieve('MUSR62262; Group; top; Asymmetry; MA')
MUSR62263__Group__top__Asymmetry__MA = ADS.retrieve('MUSR62263; Group; top; Asymmetry; MA')
MUSR62260__Group__fwd__Asymmetry__MA = ADS.retrieve('MUSR62260; Group; fwd; Asymmetry; MA')
MUSR62261__Group__fwd__Asymmetry__MA = ADS.retrieve('MUSR62261; Group; fwd; Asymmetry; MA')
MUSR62263__Group__bkwd__Asymmetry__MA = ADS.retrieve('MUSR62263; Group; bkwd; Asymmetry; MA')
MUSR62260__Group__bkwd__Asymmetry__MA = ADS.retrieve('MUSR62260; Group; bkwd; Asymmetry; MA')
MUSR62260__Group__bottom__Asymmetry__MA = ADS.retrieve('MUSR62260; Group; bottom; Asymmetry; MA')
MUSR62262__Group__bottom__Asymmetry__MA = ADS.retrieve('MUSR62262; Group; bottom; Asymmetry; MA')
MUSR62263__Group__fwd__Asymmetry__MA = ADS.retrieve('MUSR62263; Group; fwd; Asymmetry; MA')
MUSR62263__Group__bottom__Asymmetry__MA = ADS.retrieve('MUSR62263; Group; bottom; Asymmetry; MA')
MUSR62262__Group__fwd__Asymmetry__MA = ADS.retrieve('MUSR62262; Group; fwd; Asymmetry; MA')
MUSR62261__Group__bkwd__Asymmetry__MA = ADS.retrieve('MUSR62261; Group; bkwd; Asymmetry; MA')
MUSR62260__Group__top__Asymmetry__MA = ADS.retrieve('MUSR62260; Group; top; Asymmetry; MA')

def create_plot(num):
    fig, axes = plt.subplots(edgecolor='#ffffff', figsize=[8.7286, 4.9762], ncols=2, nrows=2, num=num, subplot_kw={'projection': 'mantid'})
    axes[0][0].plot(MUSR62260__Group__bottom__Asymmetry__MA, color='#1f77b4', distribution=True, label='MUSR62260; Group; bottom; Asymmetry; MA: spec 1', specNum=1)
    axes[0][0].plot(MUSR62261__Group__bottom__Asymmetry__MA, color='#ff7f0e', distribution=True, label='MUSR62261; Group; bottom; Asymmetry; MA: spec 1', specNum=1)
    axes[0][0].plot(MUSR62262__Group__bottom__Asymmetry__MA, color='#2ca02c', distribution=True, label='MUSR62262; Group; bottom; Asymmetry; MA: spec 1', specNum=1)
    axes[0][0].plot(MUSR62263__Group__bottom__Asymmetry__MA, color='#d62728', distribution=True, label='MUSR62263; Group; bottom; Asymmetry; MA: spec 1', specNum=1)
    axes[0][0].set_xlabel('Time ($\\mu s$)')
    axes[0][0].set_ylabel('Asymmetry')
    axes[0][0].set_xlim([0.0, 15.0])
    axes[0][0].set_ylim([-0.5532, 0.5774])
    axes[0][0].legend().set_draggable(True)

    axes[0][1].plot(MUSR62260__Group__bkwd__Asymmetry__MA, color='#1f77b4', distribution=True, label='MUSR62260; Group; bkwd; Asymmetry; MA: spec 1', specNum=1)
    axes[0][1].plot(MUSR62261__Group__bkwd__Asymmetry__MA, color='#ff7f0e', distribution=True, label='MUSR62261; Group; bkwd; Asymmetry; MA: spec 1', specNum=1)
    axes[0][1].plot(MUSR62262__Group__bkwd__Asymmetry__MA, color='#2ca02c', distribution=True, label='MUSR62262; Group; bkwd; Asymmetry; MA: spec 1', specNum=1)
    axes[0][1].plot(MUSR62263__Group__bkwd__Asymmetry__MA, color='#d62728', distribution=True, label='MUSR62263; Group; bkwd; Asymmetry; MA: spec 1', specNum=1)
    axes[0][1].set_xlabel('Time ($\\mu s$)')
    axes[0][1].set_ylabel('Asymmetry')
    axes[0][1].set_xlim([0.0, 15.0])
    axes[0][1].set_ylim([-0.5532, 0.5774])
    axes[0][1].legend().set_draggable(True)

    axes[1][0].plot(MUSR62260__Group__top__Asymmetry__MA, color='#1f77b4', distribution=True, label='MUSR62260; Group; top; Asymmetry; MA: spec 1', specNum=1)
    axes[1][0].plot(MUSR62261__Group__top__Asymmetry__MA, color='#ff7f0e', distribution=True, label='MUSR62261; Group; top; Asymmetry; MA: spec 1', specNum=1)
    axes[1][0].plot(MUSR62262__Group__top__Asymmetry__MA, color='#2ca02c', distribution=True, label='MUSR62262; Group; top; Asymmetry; MA: spec 1', specNum=1)
    axes[1][0].plot(MUSR62263__Group__top__Asymmetry__MA, color='#d62728', distribution=True, label='MUSR62263; Group; top; Asymmetry; MA: spec 1', specNum=1)
    axes[1][0].set_xlabel('Time ($\\mu s$)')
    axes[1][0].set_ylabel('Asymmetry')
    axes[1][0].set_xlim([0.0, 15.0])
    axes[1][0].set_ylim([-0.5532, 0.5774])
    axes[1][0].legend().set_draggable(True)

    axes[1][1].plot(MUSR62260__Group__fwd__Asymmetry__MA, color='#1f77b4', distribution=True, label='MUSR62260; Group; fwd; Asymmetry; MA: spec 1', specNum=1)
    axes[1][1].plot(MUSR62261__Group__fwd__Asymmetry__MA, color='#ff7f0e', distribution=True, label='MUSR62261; Group; fwd; Asymmetry; MA: spec 1', specNum=1)
    axes[1][1].plot(MUSR62262__Group__fwd__Asymmetry__MA, color='#2ca02c', distribution=True, label='MUSR62262; Group; fwd; Asymmetry; MA: spec 1', specNum=1)
    axes[1][1].plot(MUSR62263__Group__fwd__Asymmetry__MA, color='#d62728', distribution=True, label='MUSR62263; Group; fwd; Asymmetry; MA: spec 1', specNum=1)
    axes[1][1].set_xlabel('Time ($\\mu s$)')
    axes[1][1].set_ylabel('Asymmetry')
    axes[1][1].set_xlim([0.0, 15.0])
    axes[1][1].set_ylim([-0.5532, 0.5774])
    axes[1][1].legend().set_draggable(True)
    
    return fig, axes

time1 = time.time()
fig, axes = create_plot('Figure Reset One Axis')
plt.show()
time2 = time.time()
print('Initial plot creation time {}'.format(time2-time1))
# plt.close()
# 
# fig, axes = create_plot()
# plt.show()
time1 = time.time()
ax = fig.axes[0]
ax.cla()
ax.tracked_workspaces.clear()
ax.set_prop_cycle(None)
axes[0][0].plot(MUSR62261__Group__bottom__Asymmetry__MA, color='#ff7f0e', distribution=True, label='MUSR62261; Group; bottom; Asymmetry; MA: spec 1', specNum=1)
axes[0][0].plot(MUSR62262__Group__bottom__Asymmetry__MA, color='#2ca02c', distribution=True, label='MUSR62262; Group; bottom; Asymmetry; MA: spec 1', specNum=1)
axes[0][0].plot(MUSR62263__Group__bottom__Asymmetry__MA, color='#d62728', distribution=True, label='MUSR62263; Group; bottom; Asymmetry; MA: spec 1', specNum=1)
axes[0][0].set_xlabel('Time ($\\mu s$)')
axes[0][0].set_ylabel('Asymmetry')
axes[0][0].set_xlim([0.0, 15.0])
axes[0][0].set_ylim([-0.5532, 0.5774])
axes[0][0].legend().set_draggable(True)
time3 = time.time()
fig.canvas.draw()
time2 = time.time()
print('Time to redraw with full reset {}'.format(time2-time1))
print('Full reset just redraw time {}'.format(time2-time3))
# plt.close()

fig, axes = create_plot('Figure Remove')
plt.show()

time1 = time.time()
fig.axes[0].remove_workspace_artists(MUSR62260__Group__bottom__Asymmetry__MA)
fig.canvas.draw()
time2 = time.time()
print('Remove redraw costs {}'.format(time2-time1))
# plt.close()

fig, axes = create_plot('Figure Redraw')
plt.show()

time1 = time.time()
fig.canvas.draw()
time2 = time.time()
print('Just redraw costs {}'.format(time2-time1))
time1 = time.time()
fig.canvas.draw()
time2 = time.time()
print('Second Just redraw costs {}'.format(time2-time1))
time1 = time.time()
fig.canvas.draw()
time2 = time.time()
print('Third Just redraw costs {}'.format(time2-time1))
time1 = time.time()
fig.canvas.draw()
time2 = time.time()
print('Fourth Just redraw costs {}'.format(time2-time1))

fig, axes = create_plot('Figure Reset All Axis')
plt.show()
time1 = time.time()
for ax in fig.axes: 
    ax.cla()
    ax.tracked_workspaces.clear()
    ax.set_prop_cycle(None)
axes[0][0].plot(MUSR62260__Group__bottom__Asymmetry__MA, color='#1f77b4', distribution=True, label='MUSR62260; Group; bottom; Asymmetry; MA: spec 1', specNum=1)
axes[0][0].plot(MUSR62261__Group__bottom__Asymmetry__MA, color='#ff7f0e', distribution=True, label='MUSR62261; Group; bottom; Asymmetry; MA: spec 1', specNum=1)
axes[0][0].plot(MUSR62262__Group__bottom__Asymmetry__MA, color='#2ca02c', distribution=True, label='MUSR62262; Group; bottom; Asymmetry; MA: spec 1', specNum=1)
axes[0][0].plot(MUSR62263__Group__bottom__Asymmetry__MA, color='#d62728', distribution=True, label='MUSR62263; Group; bottom; Asymmetry; MA: spec 1', specNum=1)
axes[0][0].set_xlabel('Time ($\\mu s$)')
axes[0][0].set_ylabel('Asymmetry')
axes[0][0].set_xlim([0.0, 15.0])
axes[0][0].set_ylim([-0.5532, 0.5774])
axes[0][0].legend().set_draggable(True)

axes[0][1].plot(MUSR62260__Group__bkwd__Asymmetry__MA, color='#1f77b4', distribution=True, label='MUSR62260; Group; bkwd; Asymmetry; MA: spec 1', specNum=1)
axes[0][1].plot(MUSR62261__Group__bkwd__Asymmetry__MA, color='#ff7f0e', distribution=True, label='MUSR62261; Group; bkwd; Asymmetry; MA: spec 1', specNum=1)
axes[0][1].plot(MUSR62262__Group__bkwd__Asymmetry__MA, color='#2ca02c', distribution=True, label='MUSR62262; Group; bkwd; Asymmetry; MA: spec 1', specNum=1)
axes[0][1].plot(MUSR62263__Group__bkwd__Asymmetry__MA, color='#d62728', distribution=True, label='MUSR62263; Group; bkwd; Asymmetry; MA: spec 1', specNum=1)
axes[0][1].set_xlabel('Time ($\\mu s$)')
axes[0][1].set_ylabel('Asymmetry')
axes[0][1].set_xlim([0.0, 15.0])
axes[0][1].set_ylim([-0.5532, 0.5774])
axes[0][1].legend().set_draggable(True)

axes[1][0].plot(MUSR62260__Group__top__Asymmetry__MA, color='#1f77b4', distribution=True, label='MUSR62260; Group; top; Asymmetry; MA: spec 1', specNum=1)
axes[1][0].plot(MUSR62261__Group__top__Asymmetry__MA, color='#ff7f0e', distribution=True, label='MUSR62261; Group; top; Asymmetry; MA: spec 1', specNum=1)
axes[1][0].plot(MUSR62262__Group__top__Asymmetry__MA, color='#2ca02c', distribution=True, label='MUSR62262; Group; top; Asymmetry; MA: spec 1', specNum=1)
axes[1][0].plot(MUSR62263__Group__top__Asymmetry__MA, color='#d62728', distribution=True, label='MUSR62263; Group; top; Asymmetry; MA: spec 1', specNum=1)
axes[1][0].set_xlabel('Time ($\\mu s$)')
axes[1][0].set_ylabel('Asymmetry')
axes[1][0].set_xlim([0.0, 15.0])
axes[1][0].set_ylim([-0.5532, 0.5774])
axes[1][0].legend().set_draggable(True)

axes[1][1].plot(MUSR62260__Group__fwd__Asymmetry__MA, color='#1f77b4', distribution=True, label='MUSR62260; Group; fwd; Asymmetry; MA: spec 1', specNum=1)
axes[1][1].plot(MUSR62261__Group__fwd__Asymmetry__MA, color='#ff7f0e', distribution=True, label='MUSR62261; Group; fwd; Asymmetry; MA: spec 1', specNum=1)
axes[1][1].plot(MUSR62262__Group__fwd__Asymmetry__MA, color='#2ca02c', distribution=True, label='MUSR62262; Group; fwd; Asymmetry; MA: spec 1', specNum=1)
axes[1][1].plot(MUSR62263__Group__fwd__Asymmetry__MA, color='#d62728', distribution=True, label='MUSR62263; Group; fwd; Asymmetry; MA: spec 1', specNum=1)
axes[1][1].set_xlabel('Time ($\\mu s$)')
axes[1][1].set_ylabel('Asymmetry')
axes[1][1].set_xlim([0.0, 15.0])
axes[1][1].set_ylim([-0.5532, 0.5774])
axes[1][1].legend().set_draggable(True)  
time3 = time.time()
fig.canvas.draw()
time2 = time.time()
print('Reset all axes redraw costs {}'.format(time2-time1))
print('Reset all axes redraw only costs {}'.format(time2-time3))
```